fix meta() calls for non-storage tensors (i.e. xla)

bdhirsh opened the pull request:
It turns out that TensorIteratorBase `meta()` calls don't work with XLA tensors, since the logic that builds up the `TensorIteratorBase` object also tries to grab/store the underlying tensors' data pointers. This doesn't work for XLA because they don't have storage.


I think it's fine to just skip this bit of logic for tensors that don't have storage- since the data_ptr information isn't important to the meta call, and TensorIterator isn't actually used in the implementation for non-native kernels, i.e. XLA.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56307 fix RegistrationDeclarations.yaml, now that we codegen composite kernels for structured functional/inplace ops
* **#56306 fix meta() calls for non-storage tensors (i.e. xla)**

Differential Revision: [D27883949](https://our.internmc.facebook.com/intern/diff/D27883949)

bdhirsh committed:
fix meta() calls for non-storage tensors (i.e. xla)

[ghstack-poisoned]

bdhirsh committed:
Update on "fix meta() calls for non-storage tensors (i.e. xla)"

It turns out that TensorIteratorBase `meta()` calls don't work with XLA tensors, since the logic that builds up the `TensorIteratorBase` object also tries to grab/store the underlying tensors' data pointers. This doesn't work for XLA because they don't have storage.


I think it's fine to just skip this bit of logic for tensors that don't have storage- since the data_ptr information isn't important to the meta call, and TensorIterator isn't actually used in the implementation for non-native kernels, i.e. XLA.




[ghstack-poisoned]

ezyang reviewed (approved):
Comments? Also, I think I might prefer it if you explicitly test before the for loop if your are XLA, as the current formulation implies that it is OK for some tensors to have storage and others not, which is not actually true.

ezyang commented:
> My thought was that we want the logic to be resilient to mixed XLA/CPU inputs. Is that something that we should expect never to happen?

That's fine, but what will have happened is that we will have concluded that the common device is XLA (the same way if you have mixed CPU/CUDA, you conclude that it's CUDA), so just test on  *that*.

bdhirsh committed:
Update on "fix meta() calls for non-storage tensors (i.e. xla)"

It turns out that TensorIteratorBase `meta()` calls don't work with XLA tensors, since the logic that builds up the `TensorIteratorBase` object also tries to grab/store the underlying tensors' data pointers. This doesn't work for XLA because they don't have storage.


I think it's fine to just skip this bit of logic for tensors that don't have storage- since the data_ptr information isn't important to the meta call, and TensorIterator isn't actually used in the implementation for non-native kernels, i.e. XLA.




[ghstack-poisoned]

bdhirsh commented:
Turns out that we don't store the common device as a field on `TensorIterator`, so I added one. That's going to hurt perf just a bit (an extra field that needs to be initialized), but on the whole that shouldn't be any more expensive than what I had initially- an extra storage check on every input and output tensor.

bdhirsh committed:
Update on "fix meta() calls for non-storage tensors (i.e. xla)"

It turns out that TensorIteratorBase `meta()` calls don't work with XLA tensors, since the logic that builds up the `TensorIteratorBase` object also tries to grab/store the underlying tensors' data pointers. This doesn't work for XLA because they don't have storage.


I think it's fine to just skip this bit of logic for tensors that don't have storage- since the data_ptr information isn't important to the meta call, and TensorIterator isn't actually used in the implementation for non-native kernels, i.e. XLA.


Differential Revision: [D27883949](https://our.internmc.facebook.com/intern/diff/D27883949)

[ghstack-poisoned]
